File: packages/wasmvm/wasmlib/src/wasmtypes/scint16.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_INT16_LENGTH: usize = 2;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn int16_decode(dec: &mut WasmDecoder) -> i16 {
    dec.vli_decode(16) as i16
}

pub fn int16_encode(enc: &mut WasmEncoder, value: i16) {
    enc.vli_encode(value as i64);
}

pub fn int16_from_bytes(buf: &[u8]) -> i16 {
    if buf.len() == 0 {
        return 0;
    }
    if buf.len() != SC_INT16_LENGTH {
        panic("invalid Int16 length");
    }
    i16::from_le_bytes(buf.try_into().expect("WTF?"))
}

pub fn int16_to_bytes(value: i16) -> Vec<u8> {
    value.to_le_bytes().to_vec()
}

pub fn int16_from_string(value: &str) -> i16 {
    value.parse::<i16>().unwrap()
}

pub fn int16_to_string(value: i16) -> String {
    value.to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableInt16 {
    proxy: Proxy,
}

impl ScImmutableInt16 {
    pub fn new(proxy: Proxy) -> ScImmutableInt16 {
        ScImmutableInt16 { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        int16_to_string(self.value())
    }

    pub fn value(&self) -> i16 {
        int16_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable i16 in host container
pub struct ScMutableInt16 {
    proxy: Proxy,
}

impl ScMutableInt16 {
    pub fn new(proxy: Proxy) -> ScMutableInt16 {
        ScMutableInt16 { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: i16) {
        self.proxy.set(&int16_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        int16_to_string(self.value())
    }

    pub fn value(&self) -> i16 {
        int16_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scint16.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from '../sandbox';
import {intFromString, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';

export const ScInt16Length = 2;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function int16Decode(dec: WasmDecoder): i16 {
    return dec.vliDecode(16) as i16;
}

export function int16Encode(enc: WasmEncoder, value: i16): void {
    enc.vliEncode(value as i64);
}

export function int16FromBytes(buf: Uint8Array): i16 {
    if (buf.length == 0) {
        return 0;
    }
    if (buf.length != ScInt16Length) {
        panic('invalid Int16 length');
    }
    const ret: i16 = buf[1];
    return (ret << 8) | buf[0];
}

export function int16ToBytes(value: i16): Uint8Array {
    const buf = new Uint8Array(ScInt16Length);
    buf[0] = value as u8;
    buf[1] = (value >> 8) as u8;
    return buf;
}

export function int16FromString(value: string): i16 {
    return intFromString(value, 16) as i16;
}

export function int16ToString(value: i16): string {
    return value.toString();
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableInt16 {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return int16ToString(this.value());
    }

    value(): i16 {
        return int16FromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableInt16 extends ScImmutableInt16 {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: i16): void {
        this.proxy.set(int16ToBytes(value));
    }
}
